mbrl/util/math.py

```diff
@@ -299,3 +301,89 @@ def propagate(
     if propagation_method == "expectation":
         return propagate_expectation(predictions)
     raise ValueError(f"Invalid propagation method {propagation_method}.")
+
+
+# ------------------------------------------------------------------------ #
+# Colored noise generator for iCEM
+# ------------------------------------------------------------------------ #
+# Generate colored noise (Gaussian distributed noise with a power law spectrum)
+# Adapted from colorednoise package, credit: https://github.com/felixpatzelt/colorednoise
+def powerlaw_psd_gaussian(
+    exponent: float,
+    size: Union[int, Iterable[int]],
+    device: torch.device,
+    fmin: float = 0,
+):
+    """Gaussian (1/f)**beta noise.
+
+    Based on the algorithm in: Timmer, J. and Koenig, M.:On generating power law noise.
+    Astron. Astrophys. 300, 707-710 (1995)
+
+    Normalised to unit variance
+
+    Args:
+        exponent (float): the power-spectrum of the generated noise is proportional to
+            S(f) = (1 / f)**exponent.
+        size (int or iterable): the output shape and the desired power spectrum is in the last
+            coordinate.
+        device (torch.device): device where computations will be performed.
+        fmin (float): low-frequency cutoff. Default: 0 corresponds to original paper.
+
+    Returns
+        (torch.Tensor): The samples.
+    """
+
+    # Make sure size is a list so we can iterate it and assign to it.
+    if isinstance(size, int):
+        size = [size]
+    else:
+        size = list(size)
+
+    # The number of samples in each time series
+    samples = size[-1]
+
+    # Calculate Frequencies (we assume a sample rate of one)
+    # Use fft functions for real output (-> hermitian spectrum)
+    f = rfftfreq(samples, device=device)
```

Comment:
This uses `torch>=1.8`, which is not the version tested in the release repo, so some users might not be at this version yet.  Can we change this for a local wrapper that checks `torch`'s version and if less than 1.8 uses `numpy.fft.rfftfreq` instead?